src/core/shape_inference/include/generate_proposals_shape_inference.hpp

```diff
@@ -22,106 +22,120 @@ std::vector<TRShape> shape_infer(const GenerateProposals* op, const std::vector<
     const auto& scores_shape = input_shapes[3];
     const auto im_info_shape_rank = im_info_shape.rank();
     auto num_batches = Dimension::dynamic();
```

Comment:
Currently it is possible to build GenerateProposals shape_infer only with PartialShape class, because `num_rois` dimension is always dynamic by design. For other shape types compilation error is better than runtime exception, so incorrect usage can be detected earlier.
It can be refactored to make it non template function. To be discussed.